[1, 2단계 - 체스] 프람(김민종) 미션 제출합니다.

koust6u opened the pull request:
안녕하세요! 썬 체스 과제를 통해 찾아뵙게 된 프람입니다 .😊

이번 미션를 진행하면서 빠듯한 일정 때문에 못하게 모든 요구사항을 충족하지 못했어요.😭

부족한 부분은 따로 README파일에 정리를 두었습니다. 피드백 주신 후에 바로 해당 부분도 같이 리팩토링하도록 하겠습니다 ㅠㅠ.

잘 부탁드립니다!!

youngsu5582 committed:
docs(README) : 기능 요구사항 초안 정리

youngsu5582 committed:
feat(Point) : 위치를 가지는 도메인 생성 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Piece) : 기물 도메인 생성 기능 구현

상위 추상 클래스 Piece 아래
Bishop , King , Pawn , Knight , Queen , Rook 추가
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
refactor(Point) : 위치 도메인 기물 패키지로 이동

youngsu5582 committed:
refactor(Piece) : 기물에 색깔 추가

youngsu5582 committed:
feat(Pieces) : 기물 일급 컬렉션 생성 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(ChessBoard) : 체스판 생성 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Piece) : 비숍,킹,나이트,폰,룩,퀸 도메인 기물 상속 추가

youngsu5582 committed:
refactor(All) : 도메인들 도메인 패키지로 이동

youngsu5582 committed:
feat(ChessBoardGenerator) : 기본 체스판을 생성해주는 클래스 구현

youngsu5582 committed:
feat(ChessBoard,Pieces) : 포인트를 통해 기물을 찾는 기능 구현

Piece VO화
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(PieceStatus) : 자신의 상태를 알려주는 기능 추가

Queen,Rook,King,Pawn,Bishop,Knight 에도 반영

youngsu5582 committed:
test(ChesBoard) : 기본 체스판을 생성하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(BoardDto,BoardMapper) : DTO 변환 기능 구현

Piece 도메인에 getPoint
Point 도메인에 getFileIndex,getRankIndex
ChessBoard 도메인에 toMap 추가

youngsu5582 committed:
feat(InputView,ChessCommand) : 체스 명령어를 입력 받는 기능 구현

youngsu5582 committed:
feat(OutputView) : 체스판을 출력하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(All) : 1단계 기능 요구사항 완료

- 이에 따른 기능 요구사항 체크

youngsu5582 committed:
feat(Point) : 파일,랭크 를 포함한 문자열을 통해 생성 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Point) : 파일,랭크 를 포함한 문자열을 통해 생성 기능 구현

File,Rank 에 from 정적 팩토리 메소드 추가
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(File,Rank) : 그전 과 다음 요소 받는 기능 구현 , 경계선 위치하는지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Direction) : 방향에 따라 이동하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Pawn) : 폰이 이동하는 가능한지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Index) : 포인트 를 숫자로 변환하는 기능 구현

Direction 의 move 테스트도 변경
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(DirectionFactory) : 방향 결정해주는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
refactor(all) : 도메인 패키지 재 분리

youngsu5582 committed:
feat(Rook) : 룩 기물이 이동 가능한지 확인하는 기능 구현

youngsu5582 committed:
feat(Bishop) : 비숍 기물이 이동 가능한지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Queen) : 퀸 기물이 이동 가능한지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Index) : 인덱스가 방향에 따라 이동하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Direction) : 방향이 세로(위,아래) 인지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Pawn) : 폰이 이동하지 않을 시 두 칸 이동하는 기능 구현

Piece 멤버 변수 protected 로 변경
Point 가 방향 결정 담당하게 변경
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(King) : 킹이 이동 가능한지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Knight) : 나이트가 이동 가능한지 확인 하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Pieces) : 기물과 포인트에 있는 기물이 아군인지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Pieces) : 포인트 와 포인트 전까지 기물이 존재하는지 확인하는 기능 구현

인덱스가 해당 인덱스까지 경로를 반환하는 기능 구현
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Piece) : 기물이 흑인지 백인지 알려주는 기능 구현

youngsu5582 committed:
docs(README) : 2단계 기능 요구사항 작성

youngsu5582 committed:
feat(Movable) : 이동 가능한지 알려주는 canMove 인터페이스 추상화

기물 클래스에 반영
PieceImpl 클래스에도 반영

youngsu5582 committed:
refactor(piece) : 방향 정적 임포트로 수정

Bishop,Rook,Queen,Knight,Pawn 클래스에 반영

youngsu5582 committed:
feat(File,Rank) : 파일 , 랭크 각각 원시값을 통해 생성 구현

- 이에 따른 테스트 코드 구현

youngsu5582 committed:
feat(Pieces) : 기물이 기물들 사이 이동 가능한지 기능 구현

Direction 이 직선인지 대각선인지 확인하는 기능 추가
- 이에 따른 테스트 코드 구현

youngsu5582 committed:
feat(Pieces) : 체스기물이 이동하는 기능 구현

- 이에 따른 테스트 코드 구현

youngsu5582 committed:
feat(Pieces) : 체스기물이 이동하는 기능 구현

입출력용 RouteDTO 추가
- 이에 따른 테스트 코드 구현

youngsu5582 committed:
feat(ChessController) : 체스 게임을 진행하는 기능 구현

ChessCommand 에 MOVE 명령어 추가
- 기능 완료에 따른 요구 사항 체크

youngsu5582 committed:
refactor(Index) : list -> pathIndexes 로 변수명 변경

ChessCommand 에 MOVE 명령어 추가
- 기능 완료에 따른 요구 사항 체크

youngsu5582 committed:
refactor(All) : 도메인 패키지 재분리

youngsu5582 committed:
refactor(DirectionFactory) : 클래스 명 변경

기존 : DirectionFactory
변경 : DirectionUtil

youngsu5582 committed:
refactor(All) : 매개변수 , 지역변수 final 키워드 추가

youngsu5582 committed:
docs(README) : TODO 추가

- 도메인적 추가 구현 사항
- 도메인 외적 추가 기능 구현 사항

syoun602 commented on src/main/java/controller/ChessController.java:
```diff
@@ -0,0 +1,35 @@
+
+public class ChessController {
+
+    public void start() {
```
우선 동작이 제대로 되지 않아서 세세히 볼 수가 없네요 😢
동작이 되도록 구현해봅시다~

koust6u replied:
앗.. 죄송합니다. PR전 최종 확인을 했어야했는데 꼼꼼하게 확인 못했습니다! 코드가 돌아가도록 수정하겠습니다.

syoun602 commented on src/main/java/domain/ChessBoard.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public Piece findPieceByPoint(final Point point) {
+        final var piece = this.pieces.findPieceWithPoint(point);
+        return piece.orElseThrow(() -> new IllegalArgumentException("해당 포인트에는 기물이 없습니다"));
```
var을 쓸때와 안쓸때의 차이가 있나요? 하나로 통일하는게 좋아보입니다~
```suggestion
        return pieces.findPieceWithPoint(point)
            .orElseThrow(() -> new IllegalArgumentException("해당 포인트에는 기물이 없습니다"));
```

koust6u replied:
var를 쓸때와 안쓸때의 차이는 없습니다. 시간이 타이트하고 페어 프로그램을 진행하다 보니 각자 스타일대로 혼용하여 사용한 곳 많은 것 같습니다.  수정하겠습니다 😊

syoun602 commented on src/main/java/domain/ChessBoard.java:
```diff
@@ -0,0 +1,42 @@
+        final Point endPoint = dto.getEndPoint();
+        final Piece piece = findPieceByPoint(startPoint);
+
+        if (pieces.check(piece, endPoint)) {
```
어떤걸 check하는건가요~?

koust6u replied:
chek라는 것은 이동 첫 번째 파라미터인 pIece가 두번째 파라미터인 point로 이동이 가능한지 확인하는 기능이었습니다.

해당 기능을하는 Pieces의 check 메서드는canReplace로 이름을 변경해주어 해당 기물이 해당 위치로 대체 될 수 있는지 더 직관적인 이름으로 변경하였습니다.

syoun602 commented on src/main/java/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,76 @@
+        return value.size();
+    }
+
+    public Map<Point, Piece> toMap() {
```
애초부터 맵으로 가지고 있는건 어떤가요??

koust6u replied:
애초에 맵을 가지고 있었다면 불필요한 코드로 보입니다. 

피드백 후 제출한 코드에는 Map의 형태가 아닌 PieceDto를 리스트 형태로 가지고있도록 dto의 구조를 수정해주었습니다 😊

syoun602 commented on src/main/java/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,76 @@
+        if (!piece.canMove(endPoint)) {
+            return false;
+        }
+        return switch (piece.getStatus()) {
```
`else 예약어를 쓰지 않는다.
else 예약어를 쓰지 말라고 하니 switch/case로 구현하는 경우가 있는데 switch/case도 허용하지 않는다.`

koust6u replied:
네 최초 리뷰요청에서 시간이 많이 부족하여 요구사항보다 일단은 돌아가게 만들자 생각하여 요구사항을 어기는 코드가 상당히 많았습니다. 이번 요청에서는 하나씩 모두 수정하였습니다!

syoun602 commented on src/main/java/domain/piece/attribute/point/Direction.java:
```diff
@@ -0,0 +1,33 @@
+        return function.apply(index);
+    }
+
+    // TODO : PAWN 에서 사용하기 위함 + 함수명이 Straight 일시 왼쪽 오른쪽 선도 생각 가능
```
주석은 제거해주세요~

syoun602 commented on src/main/java/domain/piece/attribute/point/File.java:
```diff
@@ -0,0 +1,35 @@
+        this.value = value;
+    }
+
+    public static File from(final char value) {
```
간단한 stream으로 변경해봅시다

syoun602 commented on src/main/java/domain/piece/kind/Knight.java:
```diff
@@ -0,0 +1,50 @@
+import static domain.piece.attribute.point.Direction.*;
+
+public class Knight extends Piece {
+    private static final List<List<Direction>> directionList = List.of(List.of(UP, UP_RIGHT),
```
나이트의 방향을 추가하면 List<List<>>가 필요없겠네요 👀

koust6u replied:
네 좋은 아이디어 감사합니다😊  Direction(Movement로 이름 변경됨)에 나이트의 이동 방향을 추가하여 해당 List<List<>>를 대체해주었습니다

syoun602 commented on src/main/java/domain/piece/kind/Pawn.java:
```diff
@@ -0,0 +1,65 @@
+    private static final int DOUBLE_COUNT = 2;
+    private static final int SINGLE_COUNT = 1;
+
+    private boolean flag = false;
```
변수가 유의미하게 사용되고 있지 않는 것 같아요

koust6u replied:
네 사용되지 않고 있거나 무의미한 상수 역시 코드 전반적으로 모두 삭제해주었습니다!

syoun602 commented on src/main/java/domain/piece/kind/Pawn.java:
```diff
@@ -0,0 +1,65 @@
+import static domain.piece.attribute.point.Direction.*;
+
+public class Pawn extends Piece {
+    private static List<Direction> blackList = List.of(DOWN, DOWN_LEFT, DOWN_RIGHT);
+    private static List<Direction> whiteList = List.of(UP, UP_LEFT, UP_RIGHT);
```
```suggestion
    private static final List<Direction> blackList = List.of(DOWN, DOWN_LEFT, DOWN_RIGHT);
    private static final List<Direction> whiteList = List.of(UP, UP_LEFT, UP_RIGHT);
```
final이 누락되었어요~ 또한 제가 남긴 코멘트들은 프로젝트 전체적으로 훑어봐주세요😊

syoun602 commented on src/main/java/domain/piece/kind/Queen.java:
```diff
@@ -0,0 +1,22 @@
+        return PieceStatus.QUEEN;
+    }
+
+    //TODO : 현재 방향을 계산해야만 검증 로직 가능 좀 더 생각해봐야함
+    public boolean canMove(final Point point) {
```
다음 리뷰까지는 todo는 제거하거나 모두 적용해주세요~

syoun602 commented on src/main/java/util/BoardMapper.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class BoardMapper {
```
사실 dto나 mapper를 만드는건 상관이 없다고 생각되지만, 모든 동작이 제대로 안되는 상황에서 이런 클래스들을 만드는게 의미가 있는지는 잘 모르겠습니다 😢 의존성을 분리하는 것보다는 구현이 먼저 되어야 하지 않을까요?

koust6u replied:
네 썬의 의견에 전적으로 동의합니다. 일단 돌아가는 코드를 완성하고 점진적으로 리팩토링을 하도록 하겠습니다.!!!

syoun602 commented on src/main/java/util/DirectionUtil.java:
```diff
@@ -0,0 +1,98 @@
+        throw new UnsupportedOperationException("생성할 수 없습니다.");
+    }
+
+    public static Direction determineDirection(final Point standard, final Point comparison) {
```
`함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다`

syoun602 reviewed (changes requested):
안녕하세요 프람~ 리뷰어 썬입니다 ☀️
체스 미션이 쉽지 않았을텐데 고생 많으셨습니다!
우선 동작이 제대로 되고 있지 않는 것 같은데, 먼저 요구사항을 모두 충족하는 코드를 만들어보아요~

koust6u committed:
fix(ChessBoard): 기물 이동 후 프로그램 종류 되는 버그 수정

koust6u committed:
refactor(All): 지역변수들 타입 추론(var 키워드)로 타입 통일되도록 수정

koust6u committed:
refactor(All): 변화가 불필요한 인스턴스 변수 final 키워드 추가

koust6u committed:
refactor(All): todo 주석 내용 개선 및 주석 삭제

koust6u committed:
test(All): 테스트 코드 미사용 import문 삭제

koust6u committed:
feat(File): File 이 직접 이동하는 책임을 가지도록 기능 구현

koust6u committed:
feat(Rank): Rank가 직접 이동하는 책임을 가지도록 기능 구현

koust6u committed:
feat(Point): Point가 이동동하는 책임을 가지도록 기능 구현

koust6u committed:
feat(Piece): 이동 가능한 말의 위치를 받아와서 검증하는 메서드 기능 구현

koust6u committed:
refactor(Piece): 기물 이동 메서드명 일관적으로 수정

koust6u committed:
refactor(Piece): 메서드명 컨벤션에 맞게 수정

koust6u committed:
fix(Bishop): 비숍의 이동 가능위치 구하는 로직 변경에 따른 코드 수정

- canMove를 통해 해당 Point까지 이동이 가능한지를 반환하여 사용하는 것에서 이동할 수 있는 모든 경우의 Point를 반환하도록 코드를 수정하였음 이에 따라 부수적으로 수정된 코드 일괄 커밋함.

koust6u committed:
fix(SlidingPiece): 방해 기물이 없다면 특정 방향으로 계속 이동가능한 기물 추상화

koust6u committed:
fix(Queen): 퀸 다형성 적용하여 이동 가능 경로 찾는 기능 구현

koust6u committed:
fix(JumpingPiece): 뛰어넘는 기물들을 다형성 적용하여 로직 수정

뛰어넘는 기물: 킹, 나이트.

koust6u committed:
fix(Pawn): 폰 이동가능 경로 Set으로 일괄 반환하는 기능으로 구조 변경에 따른 코드 수정

-폰: Color에 따라 BlackPawn, WhitePawn폰등으로 추상화

koust6u committed:
fix(JumpingPiece): rank 인덱스 벗어나는 접근 오류 수정

koust6u committed:
fix(Pawn): 폰 매번 두번씩 이동 가능한 오류 수정

koust6u committed:
fix(Pieces): 기물들 이동하는 로직 수정

piece 인스턴스 변수를 직접 수정하는 것에서 pieces에서 기물을 빼고 넣는 방식을 수정함

koust6u committed:
fix(Pieces): 불필요한 생성자 삭제

-이에 따른 테스트 코드 및 Chessboard, ChessBoardGenerator에서 사용되는
Pieces 생성자 대체 하여줌

koust6u committed:
fix(index): 리팩토링으로 더 이상 사용되지 않는 index 삭제

-이에 따라 의존 관계의 클래스. 메서드 모두 삭제함.

koust6u committed:
fix(ChessController): Mapper로 dto 변환하는 로직 개선

koust6u committed:
refactor(ChessCommand): for문 반복에서 Stream api 사용으로 코드 변경

인덴트 뎁스 줄이기 위해서 정적 팩토리 메서드를 기존 for문에서 stream사용으로 인덴트 뎁스 해소함.

koust6u committed:
refactor(ChessCommand): 기본 체스보드 생성로직 switch문 대체

기존 switch문에서 고정된 기물 정보들을 넣어주도록 코드 수정함

koust6u committed:
fix(Piece): 불필요한 getter 삭제

getPoint 메서드 더 이상 사용되지 않으므로 코드 삭제함.

koust6u committed:
fix(Piece): 불필요한 getter 삭제

getPoint 메서드 더 이상 사용되지 않으므로 코드 삭제함.

koust6u committed:
refactor(ChessController): 커맨드 종류에 따라 생기는 if문 코드 개선

chessCommand에 따른 수행 로직을 map에 담아 처리하도록 코드 수정함.

koust6u committed:
fix(All): 미사용 메서드 삭제

koust6u committed:
refactor(File): for문을 통한 내부 상수 검색을 stream으로 대체

koust6u committed:
refactor(Rank): for문을 통한 내부 상수 검색을 stream으로 대체

koust6u committed:
feat(Controller): 입력값에 따른 예외 발생 시 재입력 기능 구현

koust6u committed:
refactor(All): 최상위 패키지 생성

koust6u committed:
fix(Piece, Pawn): 폰이 상대 공격하지 못하는 오류 수정

-Piece내에서 이동 가능한지와 Pieces에서 가능한지를 이중적으로 확인하여 발생한 오류임

koust6u committed:
feat(View): 출력 포멧에 맞게 뷰 출력, 입력 문자열 변경

koust6u committed:
feat(ChessBoard): 턴 지정 기능 구현

-WHITE 선공, BLACK후공 그 후 번갈아가며 턴이 바뀐다.

koust6u committed:
docs(README): 기능 구현 체크리스트 점검 및 삭제

-앙파상, 캐슬링 등 특이케이스 제거 및 3,4단계 미션 요구사항은 명세는 삭제함

koust6u committed:
fix(All): 일관성을 위해 파라미터 불변으로 수정

코드 일관성을 위해 가능한 모든 파라미터에 final키워드를 붙여줌.

koust6u commented:
안녕하세요 썬! 

최초 리뷰요청이 이후 버그 수정뿐만 아니라  미션 요구사항(EX, 인덴트 뎁스제한 , 세로줄 제한)을 준수하기 위해 도메인의 구조를 전반적으로 바꾸느라 재리뷰요청이 조금 늦어졌습니다.   😭

---
## 코드 구조 변경 사항

**1. Mapper를 통해 dto넘겨주는 로직 삭제후 도메인에서 dto 생성해서 넘겨주도록 바꿨습니다.**
사실 도메인의 정보를 담은 dto를 뷰로 어떻게 넘겨주어야하는지 고민이 많이 되었습니다. 또한 변경한 코드처럼 뷰에서 사용되는 dto를 도메인에서 생성한다면 뷰의 변경에 따라 도메인에도 영향이 가는 것이 올바른 접근인지 의문이 듭니다.(하지만 코드가 간결해진다는 장점때문에 위 방법으로 변경하게되었습니다)  썬의 의견은 어떠신가요?? 

**2. 기물의 특성에 따라 추상화해주었습니다.**
체스의 기물들은 크게 2가지의 방법으로 움직인다 생각하였고 해당 방법에 따라 공통된 기능이 많다고 생각되어 기물들을 이동 방법에 따라 추상화하였습니다 .
1번째는 룩, 비숍, 퀸과 같이 이동 경로 중 장애물이 없다는 진행방향으로 계속 이동이 가능한 기물들을`SlidingPiece`로 분류하였고 

2번째로는 나머지 킹,나이트, 폰(단, 폰은 킹,나이트와 다른 경우의 수가 많아 따로 추상화하였습니다)은
`JumpingPiece`로 분류하였습니다.

**3. 특정 기물의 이동 가능한 경로 구하는 기능**
특정 기물의 이동가능한 경로를 찾기 위해서 최초 리뷰요청에서는 다소 수학적(?)으로 계산하여 어느 방향으로 진행하는지  먼저 계산(이전 코드 DirectionFactory의 기능) 을 하고 해당 방향으로 이동 가능한 위치를 일일이 찾아도록하였는데 이때의 DirectionFactory에서 엄청난 if문 지옥을 개선하지 못하는 문제점이 있어 현재 코드에서는 모든 방향 찾는 방법으로 변경해주었습니다.

## 기능 추가 구현
1. 입력에 대한 오류 발생 시 재입력 받는 기능 추가 구현
2. 체스의 턴 지정해주는 기능 추가 구현 

---
마지막으로 최초 리뷰 요청에서 돌아가지 않는 코드로 리뷰에 불편을 드려 죄송합니다 😭

syoun602 commented on src/main/java/chessgame/controller/ChessController.java:
```diff
@@ -0,0 +1,43 @@
+import chessgame.view.ChessCommand;
+import chessgame.view.OutputView;
+
+public class ChessController {
```
컨트롤러보다 더 좋은 네이밍은 없을까요?
일반적으로 컨트롤러가 게임 진행과 출력 순서까지 모두 알고 있지 않을 것 같아서요!

koust6u replied:
썬 의견에 동의합니다! 의식적으로 생각을 하지 않다보니 단지 View와 Domain을 연결시켜주어 controller로 명명했었습니다. 다른 이름으로 수정하겠습니다👍

syoun602 commented on src/main/java/chessgame/controller/ChessController.java:
```diff
@@ -0,0 +1,43 @@
+
+public class ChessController {
+
+    private final Map<ChessCommand, Consumer<ChessBoard>> commandPerformances
```
커맨드에 대한 함수를 커맨드 내부에 들고 있는 것은 어떨까요?

koust6u replied:
사실 저도 커멘드에 가지고 있는것이 논리적으로 올바른거라고 생각을 했었습니다.

다만 커멘드는 view에서 입력받아 생성되기 때문에 view의 일부라 생각했습니다. 

그런데 커멘드에서 해당 맵을 가지고 있으면 결과적으로 view가 도메인(chessBoard)에 의존적이지 않을까? 라는 생각이었습니다. 

썬의 의견은 어떤가요?

syoun602 commented on src/main/java/chessgame/controller/ChessController.java:
```diff
@@ -0,0 +1,43 @@
+            = Map.of(START, ChessBoard::reset, MOVE, chessBoard -> chessBoard.move(askRoute()));
+
+    public void start() {
+        var chessBoard = new ChessBoard();
```
요기는 final이 필요없네요~

syoun602 commented on src/main/java/chessgame/controller/ChessController.java:
```diff
@@ -0,0 +1,43 @@
+    private final Map<ChessCommand, Consumer<ChessBoard>> commandPerformances
+            = Map.of(START, ChessBoard::reset, MOVE, chessBoard -> chessBoard.move(askRoute()));
+
+    public void start() {
```
start+move로 게임을 진행하다가 다시 start가 입력되면 새로 시작되는데 의도된 사항일까요?

koust6u replied:
예 맞습니다. end라는 명령어는 게임 자체를 종료한다고 가정을 했고 start는 말 그대로 게임을 시작한다는 방향으로 접근하여 게임 진행중 start를 입력 받으면 게임을 재시작한다고 가정을 했습니다.

syoun602 commented on src/main/java/chessgame/domain/ChessBoard.java:
```diff
@@ -0,0 +1,71 @@
+
+    public ChessBoard() {
+        color = Color.WHITE;
+        this.pieces = new Pieces(Set.of());
```
```
public Pieces() {
    this(Collections.emptySet());
}
```
이런 부생성자를 만들어볼 수 있을 것 같아요

syoun602 commented on src/main/java/chessgame/controller/ChessController.java:
```diff
@@ -0,0 +1,43 @@
+
+    private ChessCommand proceed(final ChessBoard chessBoard) {
+        final var chessCommand = InputView.inputChessCommand();
+        if (chessCommand == END) {
```
```suggestion
        if (chessCommand.isEnd()) {
```
와 같이 객체에게 물어봅시다

koust6u replied:
좋은 코멘트 감사합니다 ☺️

syoun602 commented on src/main/java/chessgame/domain/ChessBoard.java:
```diff
@@ -0,0 +1,71 @@
+        return ChessBoardGenerator.createDefaultBoard();
+    }
+
+    public ChessBoardDto toDto() {
+        return new ChessBoardDto(pieces.toDto());
+    }
```
도메인 내부에서 dto 변환 로직이 있는 것은 좋은 방법 같지는 않습니다.
자동차 경주에서의 요구사항을 다시 살펴볼까요?
`MVC 패턴 기반으로 리팩터링해 view 패키지의 객체가 domain 패키지 객체에 의존할 수 있지만, domain 패키지의 객체는 view 패키지 객체에 의존하지 않도록 구현한다`
근데 현재 체스보드의 import를 살펴보면 다음과 같네요.
```
import chessgame.dto.ChessBoardDto;
import chessgame.dto.RouteDto;
```
위와 같은 요구사항이 있는 이유는 간단해요.
도메인이 자주 변화하는 view로부터 분리되어 잘 변화하지 않는 비즈니스 규칙을 지키기 위해서입니다.
대부분의 변경은 view의 영역이고 따라서 dto만 수정하면 되는데, 현재 방식에서는 도메인까지 침범해야 합니다.
dto에서 도메인을 의존하는 방향으로 변경하는게 좋아보이네요 😁

koust6u replied:
썬의 친철한 답변 정말 감사합니다. 들었던 의문에 대해서 명쾌한 이해가 되었습니다. 
3,4 단계에서는 해주신 답변을 토대로 view와 domain을 분리해보겠습니다. 
해보면서 또 어려운점이 있다면 해당 pr에서 자세하게 질문해보겠습니다 !!

syoun602 commented on src/main/java/chessgame/domain/ChessBoard.java:
```diff
@@ -0,0 +1,71 @@
+                .orElseThrow(() -> new IllegalArgumentException("해당 포인트에는 기물이 없습니다"));
+    }
+
+    public void move(final RouteDto dto) {
```
dto는 말 그대로 data transfer object로써 뷰 <-> 컨트롤러 <-> 도메인 사이 사이에서 의존성을 끊기 위해 사용해야하는데, 현재는 도메인끼리도 dto로 소통하고 있네요. 차라리 dto를 전부 제거하고 도메인만 사용하는 것은 어떨까요? 
dto와 관련된 내용은 레벨 2에서 심층적으로 다루게될텐데, 레벨1의 핵심은 도메인과 객체지향에 대한 지식과 경험을 쌓는거라고 생각해서요 🙂

syoun602 commented on src/main/java/chessgame/domain/piece/Piece.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public boolean isWhite() {
+        return this.color == Color.WHITE;
```
이런 부분들도 메시지를 보내볼 수 있겠네요
```suggestion
        return color.isWhite();
```

syoun602 commented on src/main/java/chessgame/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,57 @@
+import chessgame.domain.piece.attribute.point.Point;
+import chessgame.dto.PiecesDto;
+
+import java.util.*;
```
wildcard import 단점에 대해 한번 알아봅시다~

koust6u replied:
[와일드 카드의 문제점에 대해 알아보았습니다.](https://jiga.dev/avoiding-wildcard-imports-in-java-kotlin-with-intellij/)

사실 구글 자바 스타일 가이드 컨벤션에서 와일드카드를 사용하지 말라는 내용을 인지는 하고 있었는데 급하게 코드를 작성하다보니 의도치 않게 와일드 카드 임포트가 발생했던 것 같습니다. 
이번 리뷰어님의 코멘트 덕분에 왜 컨벤션에서 사용을 지양하는지 알 수 있었습니다. 감사합니다!

syoun602 commented on src/main/java/chessgame/domain/piece/kind/pawn/BlackPawn.java:
```diff
@@ -0,0 +1,59 @@
+        super(point, color);
+    }
+
+    @Override
+    protected Set<Movement> getMovableDirection(Pieces pieces) {
+        final var availableMovement = new HashSet<Movement>();
+
+        insertAbleToAttack(pieces, availableMovement);
+
+        if (pieces.findPieceWithPoint(point.move(NORMAL_STEP)).isPresent()) {
+            return availableMovement;
+        }
+
+        insertSpecialCase(pieces, availableMovement);
+        availableMovement.add(NORMAL_STEP);
+        return availableMovement;
+    }
```
whitePawn과 정확히 일치하는 메서드네요. 중복을 제거해볼까요?

syoun602 commented on src/main/java/chessgame/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,57 @@
+
+    public void replace(final Piece piece, final Point endPoint) {
+        final var pieceWithPoint = findPieceWithPoint(endPoint);
+        pieceWithPoint.ifPresent(values::remove);
+        values.remove(piece);
```
HashSet은 내부적으로 무엇을 들고 있을까요? 
remove 메서드도 한번 살펴봅시다~ ifPresent가 꼭 필요하지 않을 것 같아서요

koust6u replied:
HashSet은 HashMap을 맴버로 가지고 있었습니다. 그런데 질문 주신 의도가 무엇인지 잘 가늠이 가지않습니다. 혹시 더 구체적으로 설명부탁드려도 괜찮을까요??

syoun602 commented on src/main/java/chessgame/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,57 @@
+                .findAny();
+    }
+
+    public boolean isFriend(final Piece piece, final Point point) {
```
friend보다 더 좋은 단어를 고민해보아요.
ally, team, color 는 어떤가요?

koust6u replied:
맞습니다. 이번 미션을 하면서 네이밍에 신경을 많이 못썼습니다. 좋은 의견감사합니다

syoun602 commented on src/main/java/chessgame/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,57 @@
+
+import java.util.*;
+
+public class Pieces {
```
대체로 필터를 통해서 찾고 ifPresent와 같은 메서드를 사용하고 있는데, position, piece를 map으로 들고 있었다면 좀 더 구현이 한눈에 들어오지 않을까? 라는 의문이 있습니다 🤔

koust6u replied:
예 저도 같은 생각입니다! buttom up 방식으로 작은 단위부터 개발 진행하다. 논리적으로 페어와 '기물이 위치를 가지고 있는게 자연스럽지 않을까?'라는 생각에서 코드를 map없이 진행하게 되었습니다. ㅜㅜ 꽤 많이 진행하고 나서 썬이 말씀하신 것과 같은 문제점을 찾게 되었습니다. 이미 구조를 수정하기 너무 늦어 그대로 코드를 완성하고 제출하게되었습니다. 3,4 단계에서 여유가 남는다면 해당 구조로 변경해보도록 하겠습니다!

syoun602 commented on src/main/java/chessgame/domain/piece/attribute/point/File.java:
```diff
@@ -0,0 +1,57 @@
+                .orElseThrow(() -> new IllegalArgumentException(String.format("%c는 파일에 존재하지 않습니다.", value)));
+    }
+
+    public boolean isFarLeft() {
+        return ordinal() == 0;
+    }
+
+    public boolean isFarRight() {
+        return ordinal() == values().length - 1;
+    }
+
+    public boolean canMoveLeft(final int step) {
+        return ordinal() - step >= 0;
+    }
+
+    public boolean canMoveRight(final int step) {
+        return ordinal() + step < values().length;
+    }
```
이펙티브 자바에 `아이템 35 : ordinal 메서드 대신 인스턴스 필드를 사용하라` 라는 내용이 있습니다~!
[해당 글](https://kukim.tistory.com/69)을 한번 살펴볼까요?

koust6u replied:
이렇게 세세하게 쪽집게식으로 짚어주셔서 정말로 감사드립니다. 바로 반영토록하겠습니다.!!!

koust6u replied:
생각하지도 못했던 내용이었습니다. 좋은 의견 감사합니다!!

syoun602 commented on src/main/java/chessgame/domain/piece/attribute/point/Movement.java:
```diff
@@ -0,0 +1,38 @@
+    DOWN(0, -1),
+    LEFT(-1, 0),
+    RIGHT(1, 0),
+    RIGHT_UP(RIGHT.x, UP.y),
+    LEFT_UP(LEFT.x, UP.y),
+    RIGHT_DOWN(RIGHT.x, DOWN.y),
+    LEFT_DOWN(LEFT.x, DOWN.y),
+    UP_UP(0, UP.y * 2),
+    DOWN_DOWN(0, DOWN.y * 2),
+    LEFT_UP_UP(LEFT.x, UP_UP.y),
+    RIGHT_UP_UP(RIGHT.x, UP_UP.y),
+    LEFT_DOWN_DOWN(LEFT.x, DOWN_DOWN.y),
+    RIGHT_DOWN_DOWN(RIGHT.x, DOWN_DOWN.y),
+    LEFT_LEFT_UP(LEFT.x * 2, UP.y),
+    LEFT_LEFT_DOWN(LEFT.x * 2, DOWN.y),
+    RIGHT_RIGHT_UP(RIGHT.x * 2, UP.y),
+    RIGHT_RIGHT_DOWN(RIGHT.x * 2, DOWN.y);
```
모두 숫자로 통일하는 것은 어떤가요? (1, 1) (1, 2) (-1, 0) 등
현재 방식이 가독성이 좋은지 한번 의심해볼 필요가 있는 것 같아요

koust6u replied:
좋은 의견 감사합니다. 말씀 주신거처럼 숫자로 직접 넣어주는 것이 훨씬 가독성 측면에서 좋다 생각이 됩니다!

syoun602 commented on src/main/java/chessgame/domain/piece/attribute/point/Movement.java:
```diff
@@ -0,0 +1,38 @@
+        this.y = y;
+    }
+
+    int x() {
+        return x;
+    }
+
+    int y() {
+        return y;
+    }
```
Lombok 어노테이션을 쓰든, IDE에서 getter를 자동생성하든 get을 항상 붙여주기 때문에
저는 개인적으로 자바 언어를 사용할때는 단순 값 반환시 get을 붙이는 편입니다.

koust6u replied:
네 말씀 주신 코멘트를 보고 찾아보니 자바 빈즈 규약 중`클래스의 속성들은 get, set 혹은 표준 명명법을 따르는 메서드들을 사용해 접근할 수 있어야 한다.` 있다는 것을 찾을 수 있었습니다. 

규약에 따라 관련 메소드에 일관적으로 get를 붙여주는 것이 좀 더 바람직하다고 느꼈습니다. 
감사합니다😊

syoun602 commented on src/main/java/chessgame/domain/piece/kind/jumping/King.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @Override
+    protected Set<Movement> getMovableDirection() {
+        return Set.of(Movement.UP, Movement.DOWN, Movement.LEFT, Movement.RIGHT, Movement.RIGHT_UP, Movement.RIGHT_DOWN,
+                Movement.LEFT_UP, Movement.LEFT_DOWN);
+    }
```
List로도 충분해보이는데 Set을 사용하신 이유가 있을까요?

koust6u replied:
제가 알기로는 set list의 차이는 
Set은 중복을 허용하지 않고 List는 중복을 허용하지만 순서를 보장한다는 특징이 있는 것으로 압니다. 

굳이 순서에 대해 보장해줄 필요가 없고 중복 역시 필요가 없다는 것을 자료형을 통해 묵시적으로 표현하고 싶어 사용하였습니다.

혹시 위와 같이 굳이 어떤 자료구조를 사용하더라도 상관이 없을 경우 Set보다 List가 권장되는 이유가 있을까요?? 
썬의 의견이 궁금합니다!!

syoun602 reviewed (approved):
안녕하세요 프람~

체스 게임 잘 구현해주셨네요~ 전부 동작하는 것 확인했습니다 👍 
코멘트를 통해 점진적으로 살펴보려 했는데, 리뷰 반영과 별개로 스스로 리팩토링을 진행하시느라 시간이 꽤 걸렸군요..! 많은 변경이 생겨 1,2단계에서 조금 더 얘기해보고 싶은 부분들이 있는데, 지금 변경 요청을 드리면 3,4단계에 대한 부담이 생길 것 같아서 다음 단계에서 더 얘기나눠보도록 하겠습니다 🥲 

남긴 코멘트는 다음 단계에서 반영해주세요! 그럼 다음 단계도 화이팅입나다 ~ 💪

koust6u commented:
전반적으로 3,4단계 구현 시간이 빠듯할거 같아 말씀 주신 코멘트들을 적용해보았습니다. 3.4단계를 진행하면서 적용못한 코멘트들과 코멘트들에 대한 저의 의견을 첨언하여 pr하도록하겠습니다. 꼼꼼한 코드 리뷰 정말 감사합니다!!!
